File: bms/go/geo/geo_r3_Vector_Angle/geo_r3_Vector_Angle.go
package test

import (
	"math"
)

var ()

const (
	geo_s1_Radian geo_s1_Angle = 1
)

func (v geo_r3_Vector) Angle(ov geo_r3_Vector) geo_s1_Angle {
	return geo_s1_Angle(math.Atan2(v.Cross(ov).Norm(), v.Dot(ov))) * geo_s1_Radian
}

func (v geo_r3_Vector) Cross(ov geo_r3_Vector) geo_r3_Vector {
	return geo_r3_Vector{
		float64(v.Y*ov.Z) - float64(v.Z*ov.Y),
		float64(v.Z*ov.X) - float64(v.X*ov.Z),
		float64(v.X*ov.Y) - float64(v.Y*ov.X),
	}
}

func (v geo_r3_Vector) Norm() float64 { return math.Sqrt(v.Dot(v)) }

func (v geo_r3_Vector) Dot(ov geo_r3_Vector) float64 {
	return float64(v.X*ov.X) + float64(v.Y*ov.Y) + float64(v.Z*ov.Z)
}

type geo_r3_Vector struct {
	X, Y, Z float64
}

type geo_s1_Angle float64

File: translations_only/fail/go/geo/geo_r3_Vector_Angle/mistral/BaseRepair/geo_r3_Vector_Angle.rs

use std::f64;

const GEO_S1_RADIAN: f64 = 1.0;

struct GeoR3Vector {
    x: f64,
    y: f64,
    z: f64,
}

type GeoS1Angle = f64;

fn angle(v: &GeoR3Vector, ov: &GeoR3Vector) -> GeoS1Angle {
    let cross_norm = v.cross(ov).norm();
    let dot = v.dot(ov);
    GEO_S1_RADIAN * f64::atan2(cross_norm, dot)
}

impl GeoR3Vector {
    fn cross(&self, ov: &GeoR3Vector) -> GeoR3Vector {
        GeoR3Vector {
            x: self.y * ov.z - self.z * ov.y,
            y: self.z * ov.x - self.x * ov.z,
            z: self.x * ov.y - self.y * ov.x,
        }
    }

    fn norm(&self) -> f64 {
        f64::sqrt(self.dot(self))
    }

    fn dot(&self, ov: &GeoR3Vector) -> f64 {
        self.x * ov.x + self.y * ov.y + self.z * ov.z
    }
}
